### ISSUE TITLE
(core): addUnacknowledgeableWarning() method for Annotations 

### ISSUE THREAD

**Issue by dontirun** [aws/aws-cdk#26914]
### Describe the feature

Exactly the [addWarning](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.Annotations.html#addwbrwarningmessagespan-classapi-icon-api-icon-deprecated-titlethis-api-element-is-deprecated-its-use-is-not-recommended%EF%B8%8Fspan) method, but not deprecated 

### Use Case

[cdk-nag](https://github.com/cdklabs/cdk-nag) uses the cdk Annotation system to emit `Warnings` and `Errors`. Currently it uses the [addWarning](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.Annotations.html#addwbrwarningmessagespan-classapi-icon-api-icon-deprecated-titlethis-api-element-is-deprecated-its-use-is-not-recommended%EF%B8%8Fspan) method  for `Warnings`. `addWarning` was deprecated with the [introduction of `addWarningV2`](https://github.com/aws/aws-cdk/pull/26144) in the latest cdk release. All `cdk-nag` Warnings now emit deprecation messages along with the warning, which adds a lot of clutter the the CLI output. While this could be fixed by switching to `addWarningV2`, I don't think that `addWarningV2` is the correct choice in this case, primarily due to how the `cdk-nag` warning/error suppression system is intended to function.

#### Primary Concern `acknowledgeWarning` vs `NagSuppression`

[`NagSuppression`s](https://github.com/cdklabs/cdk-nag#suppressing-a-rule) are `cdk-nag`'s way of acknowledging warnings and errors. The key difference between `acknowledgeWarning` and `NagSuppression`s are that `NagSuppresion`s write the suppression to the CloudFormation Metadata in a `cdk-nag` [specific format](https://github.com/cdklabs/cdk-nag#using-on-cloudformation-templates). That Cfn metadata is important when using SAST tooling to look over the generated CloudFormation templates (Anecdotally, I know of several use cases where the `cdk-nag` Metadata is reviewed in CI/CD pipelines).  If `cdk-nag` were to switch from `addWarning` to `addWarningV2` users would be able to use both `acknowledgeWarning` and `NagSuppresion`s to silence warnings. While technically not a breaking change giving users the ability to use `acknowledgeWarning` for `cdk-nag` isn't a good choice.

#### Smaller Concerns/Annoyances with `addWarningV2` and `cdk-nag`

if `cdk-nag` were to switch to `addWarningV2`  and we were to ignore the primary concern there would be a few other minor annoyances.

1. The CLI  output of `warnings` and `errors` would be slightly different from one another. `Warnings` would have an additional `addWarningV2` identifier appended to the end of the message, while `errors` would not. While that would be the existing `RuleId` already used for `cdk-nag`, `errors` don't have a corresponding `addErrorV2` (and in my opinion it doesn't make sense to have this) so this adds a bit of clutter and confusion
2. Since there is no `addErrorV2` method there would a delta between the options to acknowledge/suppress `Warnings` and `Errors`

### Proposed Solution

I don't think that "undeprecating" `addWarning` is the correct solution, because in  cdk internal uses cases`addWarningV2` is the correct  choice and users should use `addWarningV2`. Having a clone of `addWarning` with a better name would add some "guardrails"(?)  on usage

1. Add an `addUnacknowledgeableWarning()` method to `Annotations` which is a clone of of the current `addWarning` method
2. Put a disclaimer message message that users should likely be using `addWarningV2` unless they have a specific use case where the warnings should not be acknowledgeable and that unhandled warnings will prohibit users from deploying cdk applications with `strict` mode. 

### Other Information

Related to https://github.com/cdklabs/cdk-nag/issues/1418

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.93.0

### Environment details (OS name and version, etc.)

Not relevant, but macOs 13.4 (Ventura) 😄

**Comment by indrora** [comment]
I'm torn if this is a bug or a feature request. I want to err on the side of bug. 

I'm going to consider this a bug

**Comment by mrgrain** [comment]
Thanks @dontirun we are undeprecating the method until we have a better solution for you.

We are currently thinking something like an accessible report of acknowledged warnings for plugins to parse.
`NagSuppressions` could then be a thin wrapper around CDK functionality.
But we'd also have to address the papercuts you've mentioned.

**Comment by rix0rrr** [comment]
On the papercuts:

>  this adds a bit of clutter and confusion

I appreciate this, but I also think it's vitally important that the suppression tag is surfaced in a clear way to users, especially if we intend to go full-hog on warnings (which I want to, with all the tweakable and adjustable things and weird edge cases people can get into, it's much better to overwarn and allow it to be silenced, than to not warn or \*gasp\* error). I opted for putting the tag into the warning string directly, so there's no way to miss it. Of course, any presentation layer could be aware of this, parse out the tag and display it differently, for example in a separate table column or something.

> there would a delta [...] to acknowledge/suppress [...] Errors

Do you have an example of a suppressible error? To my mind the major distinction between warnings and errors is their suppressibility, and so the fact that they cannot be suppressed is a feature.

**Comment by dontirun** [comment]
> Do you have an example of a suppressible error? To my mind the major distinction between warnings and errors is their suppressibility, and so the fact that they cannot be suppressed is a feature.

Yes, but with some context first 😄. `cdk-nag` has a concept of `NagPack`s which are sets of guidance that are based on some set of standards. Generally during a security review/audit you provide evidence to compliance and provide/discuss/ document exceptions with the reviewer/auditor.

Given that, let's say that you are using the HIPAA security based `NagPack` and encounter the [HIPAA.Security-LambdaInsideVPC](https://docs.aws.amazon.com/config/latest/developerguide/lambda-inside-vpc.html) error. If you are strictly following the HIPAA guidelines, you want to prevent a deployment that has a misconfiguration (Error), but there are plenty of valid reasons why this particular guideline doesn't matter in a specific case. For example, maybe your lambda function is calling an external service, your application is completely serverless and doesn't have a VPC otherwise, your lambda function isn't processing any Health related data, etc. In that case you would suppress with your documented reason and continue the deployment.

**Comment by rix0rrr** [comment]
In CDK Core feature parlance, would it be equivalent to say that message is a **warning**, and everyone must always run with `--strict` ? That is, all warnings must either be fixed or acknowledged?

**Comment by dontirun** [comment]
> In CDK Core feature parlance, would it be equivalent to say that message is a **warning**, and everyone must always run with `--strict` ? That is, all warnings must either be fixed or acknowledged?

Yes, I would say an error in cdk-nag is equivalent to this
